Indexing/CodeTree.cpp

```diff
@@ -357,109 +360,137 @@ bool CodeTree::CodeOp::equalsForOpMatching(const CodeOp& o) const
-      reinterpret_cast<size_t>(this)-opOfs);
+  ASS(isSearchStruct());
+  return GET_CONTAINING_OBJECT(CodeTree::SearchStruct,landingOp,this);
+}
 
-  return res;
+std::ostream& operator<<(std::ostream& out, const CodeTree::CodeOp& op)
+{
+  switch (op.instrPrefix()) {
+  case CodeTree::SUCCESS_OR_FAIL:
+    if (op.isSuccess()) {
+      out << "success";
+    } else {
+      out << "fail";
+    }
+    break;
+  case CodeTree::LIT_END:
+    out << "lit end";
+    break;
+  case CodeTree::CHECK_GROUND_TERM:
+    out << "check ground term " << *op.getTargetTerm();
+    break;
+  case CodeTree::SUFFIX_INSTR:
+    switch(op.instrSuffix()) {
+    case CodeTree::CHECK_FUN:
+      out << "check fun " << env.signature->getFunction(op.arg())->name();
+      break;
+    case CodeTree::ASSIGN_VAR:
+      out << "assign var X" << op.arg();
+      break;
+    case CodeTree::CHECK_VAR:
+      out << "check var X" << op.arg();
+      break;
+    case CodeTree::SEARCH_STRUCT:
+      out << "search struct ";
+      auto ss = op.getSearchStruct();
+      switch(ss->kind) {
+        case CodeTree::SearchStruct::FN_STRUCT: {
+          auto fn_ss = static_cast<CodeTree::FnSearchStruct*>(ss);
+          out << "length " << fn_ss->length();
+          for (unsigned i = 0; i < fn_ss->length(); i++) {
+            out << " " << fn_ss->values[i] << " ";
+            if (fn_ss->targets[i]) {
+              out << *fn_ss->targets[i];
+            } else {
+              out << "nullptr";
+            }
+          }
+          break;
+        }
+        case CodeTree::SearchStruct::GROUND_TERM_STRUCT: {
+          auto gt_ss = static_cast<CodeTree::GroundTermSearchStruct*>(ss);
+          out << "length " << gt_ss->length();
+          for (unsigned i = 0; i < gt_ss->length(); i++) {
+            out << " " << *gt_ss->values[i] << " ";
+            if (gt_ss->targets[i]) {
+              out << *gt_ss->targets[i];
+            } else {
+              out << "nullptr";
+            }
+          }
+          break;
+        }
+      }
+      break;
+    }
+    break;
+  }
+  return out;
 }
 
-CodeTree::SearchStruct::SearchStruct(Kind kind)
+CodeTree::SearchStruct::SearchStruct(Kind kind, size_t length)
 : kind(kind)
 {
   landingOp.setAlternative(0);
   landingOp.setLongInstr(SEARCH_STRUCT);
-}
+  ASS(length);
 
-void CodeTree::SearchStruct::destroy()
-{
-  switch(kind) {
-  case FN_STRUCT:
-    delete static_cast<FnSearchStruct*>(this);
-    break;
-  case GROUND_TERM_STRUCT:
-    delete static_cast<GroundTermSearchStruct*>(this);
-    break;
-  }
+  targets.reserve(length);
 }
 
+template<bool doInsert>
 bool CodeTree::SearchStruct::getTargetOpPtr(const CodeOp& insertedOp, CodeOp**& tgt)
 {
   switch(kind) {
   case FN_STRUCT:
     if(!insertedOp.isCheckFun()) { return false; }
-    tgt=&static_cast<FnSearchStruct*>(this)->targetOp(insertedOp.arg());
+    tgt=&static_cast<FnSearchStruct*>(this)->targetOp<doInsert>(insertedOp.arg());
     return true;
```

Comment:
I'm not sure returning anything here makes sense as we only check the result in debug mode where the violation below would kick in, otherwise it's simply undefined behaviour

Reply:
If you can put the assertion "inside" somehow, do it. Otherwise I think it's fine, it indicates *something* and the compiler will happily inline/discard unused values.